test/test_cli.rb

```diff
@@ -9,34 +9,31 @@ class TestCLI < Minitest::Test
 
   def setup
     @environment = 'production'
-    @tmp_file = Tempfile.new("puma-test")
-    @tmp_path = @tmp_file.path
-    @tmp_file.close!
-
-    @tmp_path2 = "#{@tmp_path}2"
+    @tmp_path  = "tmp/#{full_file_name}"
+    @tmp_path2 = "#{@tmp_path}_2"
 
     File.unlink @tmp_path  if File.exist? @tmp_path
     File.unlink @tmp_path2 if File.exist? @tmp_path2
 
     @wait, @ready = IO.pipe
 
     @events = Puma::Events.strings
-    @events.on_booted { @ready << "!" }
+    @events.on_booted { @ready.syswrite '!' }
```

Comment:
What was wrong with the shovel?

Reply:
Sometime ago I came across info implying that mixing 'sys' & 'normal' on the same IO may cause odd issues.  Ruby version or platform specific, I don't recall, nor how long ago...